### ISSUE TITLE
access trigger with remember 

### ISSUE THREAD

**Issue by lloricode** [benedmunds/CodeIgniter-Ion-Auth#1075]
Good pm, @benedmunds.

I just want to set hook with this. 

```php
	public function set_session($user)
	{

		$this->trigger_events('pre_set_session');

		$session_data = array(
		    'identity'             => $user->{$this->identity_column},
		    $this->identity_column             => $user->{$this->identity_column},
		    'email'                => $user->email,
		    'user_id'              => $user->id, //everyone likes to overwrite id so we'll use user_id
		    'old_last_login'       => $user->last_login,
		    'last_check'           => time(),
		);

		$this->session->set_userdata($session_data);

		$this->trigger_events('post_set_session');

		return TRUE;
	}
```

it works fine though...

but when `remember` is set to `TRUE`  after login user, then reach the `user_expire`

I see that the function I want to trigger is not called

even `pre_set_session` OR `post_set_session`

Thanks for response :)

**Comment by benedmunds** [comment]
Sorry I'm not sure if I'm following exactly. 

So with a remembered user, the pre_set_session event is not triggering?  Can you add debug code to ensure it's getting to that line in the code?

Is the post_login_remembered_user event triggering?

**Comment by lloricode** [comment]
I share my test code



##
just to make sure
**index.php**

`define('DEBUG', 'CI_DEBUG ');`

##
**config/ion_auth.php**
`$config['user_expire']                = 0;   `
##
**config/config.php**
```php
$config['sess_expiration'] = 3;   //3 seconds for testting
  $config['log_threshold'] = 2;
```
##
adding log message
**Ion_auth_model.php**
```php
public function set_session($user)
{

        log_message('debug', DEBUG.'pre_set_session');//<------------------
        
        $this->trigger_events('pre_set_session');

        $session_data = array(
            'identity'             => $user->{$this->identity_column},
            $this->identity_column             => $user->{$this->identity_column},
            'email'                => $user->email,
            'user_id'              => $user->id, //everyone likes to overwrite id so we'll use user_id
            'old_last_login'       => $user->last_login,
            'last_check'           => time(),
        );

        $this->session->set_userdata($session_data);

        $this->trigger_events('post_set_session');
        
        log_message('debug', DEBUG.'post_set_session');//<------------------

        return TRUE;
}
```

##
**controller**
`print_r($this->session->userdata());`


##
**core/MY_Controller.php**
```php
function __construct()
{
        parent::__construct();
        if ($this->session->has_userdata('user_id'))
        {
                $this->ion_auth->set_hook(
                        'pre_set_session', 'set_session_data_session_event', $this/* $this because the class already extended */, 'set_session_data_session', array()
                );
        }
}

public function set_session_data_session()
{
        log_message('debug', DEBUG . 'MY FUNCTION');
        // show_error('aaaa');
        $is_dean          = FALSE;
        $dean_course_id   = NULL;
        $dean_course_code = NULL;
        if ($this->ion_auth->in_group($this->config->item('user_group_dean')))
        {
                $is_dean = TRUE;
                $this->load->model('Dean_course_model');
                $obj     = $this->Dean_course_model->where(array(
                            'user_id' => $this->ion_auth->get_user_id()
                        ))->get();
                if ($obj)
                {
                        $this->load->model('Course_model');
                        $dean_course_id   = $obj->course_id;
                        $dean_course_code = $this->Course_model->get($obj->course_id)->course_code;
                }
        }
        //set the user name/last name in session
        $user_obj = $this->ion_auth->user()->row();
        $this->session->set_userdata(array(
            'user_first_name'          => $user_obj->first_name,
            'user_last_name'           => $user_obj->last_name,
            'user_fullname'            => $user_obj->last_name . ', ' . $user_obj->first_name,
            'gen_code'                 => $user_obj->gen_code, //this will be use for checking multiple logged machines in one account
            'user_groups_descriptions' => $this->_current_group_string(),
            'user_groups_names'        => $this->_current_group_string('name'),
            'user_is_dean'             => $is_dean,
            'user_dean_course_id'      => $dean_course_id,
            'user_dean_course_code'    => $dean_course_code,
        ));
}
```

##

then here how the output:

when i log in success and with remeber checked
the `log` and `print_r` result are

log
```
DEBUG - 2017-04-24 13:07:01 --> UTF-8 Support Enabled
DEBUG - 2017-04-24 13:07:01 --> Global POST, GET and COOKIE data sanitized
DEBUG - 2017-04-24 13:07:01 --> Config file loaded: /opt/lampp/htdocs/ci-capstone/application/config/common/config.php
DEBUG - 2017-04-24 13:07:01 --> Config file loaded: /opt/lampp/htdocs/ci-capstone/application/config/common/user_groups.php
DEBUG - 2017-04-24 13:07:01 --> CI_DEBUG pre_conntruct_ion_auth_lib
DEBUG - 2017-04-24 13:07:01 --> Config file loaded: /opt/lampp/htdocs/ci-capstone/application/config/ion_auth.php
DEBUG - 2017-04-24 13:07:01 --> Session: "sess_save_path" is empty; using "session.save_path" value from php.ini.
DEBUG - 2017-04-24 13:07:01 --> CI_DEBUG post_conntruct_ion_auth_lib
DEBUG - 2017-04-24 13:07:01 --> Config file loaded: /opt/lampp/htdocs/ci-capstone/application/config/breadcrumbs.php
DEBUG - 2017-04-24 13:07:01 --> CI_DEBUG pre_set_session
DEBUG - 2017-04-24 13:07:01 --> CI_DEBUG post_set_session
DEBUG - 2017-04-24 13:07:01 --> Cache class already loaded. Second attempt ignored.
DEBUG - 2017-04-24 13:07:01 --> Cache class already loaded. Second attempt ignored.
DEBUG - 2017-04-24 13:07:01 --> CI_DEBUG MY FUNCTION
DEBUG - 2017-04-24 13:07:01 --> UTF-8 Support Enabled
DEBUG - 2017-04-24 13:07:01 --> Global POST, GET and COOKIE data sanitized
DEBUG - 2017-04-24 13:07:01 --> Config file loaded: /opt/lampp/htdocs/ci-capstone/application/config/common/config.php
DEBUG - 2017-04-24 13:07:01 --> Config file loaded: /opt/lampp/htdocs/ci-capstone/application/config/common/user_groups.php
DEBUG - 2017-04-24 13:07:01 --> CI_DEBUG pre_conntruct_ion_auth_lib
DEBUG - 2017-04-24 13:07:01 --> Config file loaded: /opt/lampp/htdocs/ci-capstone/application/config/ion_auth.php
DEBUG - 2017-04-24 13:07:01 --> Session: "sess_save_path" is empty; using "session.save_path" value from php.ini.
DEBUG - 2017-04-24 13:07:01 --> CI_DEBUG post_conntruct_ion_auth_lib
DEBUG - 2017-04-24 13:07:01 --> Config file loaded: /opt/lampp/htdocs/ci-capstone/application/config/breadcrumbs.php
DEBUG - 2017-04-24 13:07:01 --> Cache class already loaded. Second attempt ignored.
DEBUG - 2017-04-24 13:07:01 --> Cache class already loaded. Second attempt ignored.
DEBUG - 2017-04-24 13:07:01 --> Cache class already loaded. Second attempt ignored.
DEBUG - 2017-04-24 13:07:01 --> Total execution time: 0.1898

```
`print_r`
```php
Array
(
    [__ci_last_regenerate] => 1493010421
    [identity] => administrator
    [username] => administrator
    [email] => admin@admin.com
    [user_id] => 1
    [old_last_login] => 1493010089
    [last_check] => 1493010421
    [user_first_name] => Admin
    [user_last_name] => istrator
    [user_fullname] => istrator, Admin
    [gen_code] => gekEMU.hYo0NQR7XBDq7le
    [user_groups_descriptions] => Administrator
    [user_groups_names] => admin
    [user_is_dean] => 
    [user_dean_course_id] => 
    [user_dean_course_code] => 
)
```

##

then after 3 second session will expire so it will relogin by remember function

log
```
DEBUG - 2017-04-24 13:08:30 --> UTF-8 Support Enabled
DEBUG - 2017-04-24 13:08:30 --> Global POST, GET and COOKIE data sanitized
DEBUG - 2017-04-24 13:08:30 --> Config file loaded: /opt/lampp/htdocs/ci-capstone/application/config/common/config.php
DEBUG - 2017-04-24 13:08:30 --> Config file loaded: /opt/lampp/htdocs/ci-capstone/application/config/common/user_groups.php
DEBUG - 2017-04-24 13:08:30 --> CI_DEBUG pre_conntruct_ion_auth_lib
DEBUG - 2017-04-24 13:08:30 --> Config file loaded: /opt/lampp/htdocs/ci-capstone/application/config/ion_auth.php
DEBUG - 2017-04-24 13:08:30 --> Session: "sess_save_path" is empty; using "session.save_path" value from php.ini.
DEBUG - 2017-04-24 13:08:30 --> CI_DEBUG post_conntruct_ion_auth_lib
DEBUG - 2017-04-24 13:08:30 --> CI_DEBUG pre_set_session
DEBUG - 2017-04-24 13:08:30 --> CI_DEBUG post_set_session
DEBUG - 2017-04-24 13:08:30 --> CI_DEBUG post_conntruct_ion_auth_lib
DEBUG - 2017-04-24 13:08:30 --> Config file loaded: /opt/lampp/htdocs/ci-capstone/application/config/breadcrumbs.php
DEBUG - 2017-04-24 13:08:30 --> Cache class already loaded. Second attempt ignored.
DEBUG - 2017-04-24 13:08:30 --> Cache class already loaded. Second attempt ignored.
DEBUG - 2017-04-24 13:08:30 --> Cache class already loaded. Second attempt ignored.
DEBUG - 2017-04-24 13:08:30 --> Total execution time: 0.3028

```
`print_r`
```php
Array
(
    [__ci_last_regenerate] => 1493010510
    [identity] => administrator
    [username] => administrator
    [email] => admin@admin.com
    [user_id] => 1
    [old_last_login] => 1493010500
    [last_check] => 1493010510
)




```

**Comment by lloricode** [comment]
opps sorry 

I mean in first commet is 
`$config['sess_expiration']`

not `user_expire`

**Comment by lloricode** [comment]
also `post_login_remembered_user` not trigger, same result of `pre_set_session` and `post_set_session`

**Comment by benedmunds** [comment]
Im not sure if using this in the MY_controller construct will work correctly if it's autoloaded.  In your actual controller try loading ion auth (not autoload) and then set the hooks.

**Comment by lloricode** [comment]
hmmm the problem is, i have more the 20 controllers, so thats why i put it in MY_controllers,

then I remove in auto_load a ion_auth and load it manually in controller.

same result as my issue

**Comment by benedmunds** [comment]
You tried it in a specific controller or the MY_controller?  Can you make an encapsulated test project and zip it/send it or upload to GH somewhere?

**Comment by lloricode** [comment]
hmm ok sir, maybe later, i will make a test project. i will notify you sir

**Comment by lloricode** [comment]
closing and resolved from PR #1078, thanks @benedmunds and @avenirer :+1:
